core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -718,6 +718,12 @@ private[spark] class DAGScheduler(
                 // read from merged output as the MergeStatuses are not available.
                 if (!mapStage.isAvailable || !mapStage.shuffleDep.shuffleMergeFinalized) {
                   missing += mapStage
+                } else if (mapStage.nextAttemptId == 0) {
+                  // Forward the nextAttemptId if skipped and get visited for the first time.
+                  // Otherwise, once it gets retried,
+                  // 1) the stuffs in stage info become distorting, e.g. task num, input byte, e.t.c
+                  // 2) the first attempt starts from 0-idx, it will not be marked as a retry
+                  mapStage.nextAttemptId += 1
```

Comment:
this doesn't feel like the best way to do this. I also don't think we should  expose nextAttemptId here, I need to look a bit more at it and will get back.

Reply:
ok, not really seeing any better way to do this. The one thing that might be nice is instead of exposing nextAttemptId here, make like a mapStage.skip() type function which can internally check if mapStage.nextAttemptId == 0 and increment if needed.  its really not to big of a deal though as we reach into Stage a lot from the DAGScheduler.

Reply:
> its really not to big of a deal though as we reach into Stage a lot from the DAGScheduler.

+1, I prefer to minimize the diff

Reply:
https://github.com/apache/spark/blob/c4e975e175c01f67ece7ae492a79554ad1b44106/core/src/main/scala/org/apache/spark/scheduler/Stage.scala#L98-L108
Don't we automatically increase the nextAttemptId whenever submitting a new stage attempt? Why the skipped stage be an abnormal case?

Reply:
its incrementing nextAttemptId (ie the next attempt, not the current one).  The skipped attempt never calls makeNewStageAttempt so the first time it was skipped this was never incremented. The default of nextAttemptId on Stage creation is 0 and since it was never incremented in the skipped stage when you call makeNewStageAttempt  for the failed attempt it is run as 0 but the next attempt would be 1.  So the skipped attempt and the rerun failed attempt both are attempt 0.

Reply:
Agree with Tom, this is specific to the skipped -> running transition; not applicable for retry which, as you @Ngone51 mentioned, will increment via `makeNewStageAttempt`

Reply:
+1 on explicitly making intent to skip clear, instead of only bumping the attempt id.

Reply:
oh I misread it. It's for the increment is for the next stage attempt. thanks @tgravescs @mridulm